File: experiments/mathexpr-skill-python/mathexpr.py
"""
Math expression parser and evaluator.

Evaluates strings like "2 + 3 * (4 - 1)" with correct operator precedence,
parentheses, and error handling. Zero external dependencies.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Union


# ---------------------------------------------------------------------------
# token-types
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class Token:
    kind: str  # "number"|"plus"|"minus"|"star"|"slash"|"percent"|"power"|"lparen"|"rparen"
    value: str


# ---------------------------------------------------------------------------
# ast-types
# ---------------------------------------------------------------------------

@dataclass
class NumberLiteral:
    type: str
    value: float

    def __init__(self, value: float) -> None:
        self.type = "number"
        self.value = value


@dataclass
class UnaryExpr:
    type: str
    op: str
    operand: AstNode

    def __init__(self, op: str, operand: AstNode) -> None:
        self.type = "unary"
        self.op = op
        self.operand = operand


@dataclass
class BinaryExpr:
    type: str
    op: str
    left: AstNode
    right: AstNode

    def __init__(self, op: str, left: AstNode, right: AstNode) -> None:
        self.type = "binary"
        self.op = op
        self.left = left
        self.right = right


AstNode = Union[NumberLiteral, UnaryExpr, BinaryExpr]


# ---------------------------------------------------------------------------
# tokenizer
# ---------------------------------------------------------------------------

def tokenize(input_str: str) -> list[Token]:
    """Convert a math expression string into a list of tokens."""
    tokens: list[Token] = []
    i = 0
    n = len(input_str)

    while i < n:
        ch = input_str[i]

        # Skip whitespace
        if ch in (' ', '\t', '\n', '\r'):
            i += 1
            continue

        # Numbers (digits or leading dot)
        if ch.isdigit() or ch == '.':
            start = i
            has_dot = False
            if ch == '.':
                has_dot = True
            i += 1
            while i < n and (input_str[i].isdigit() or input_str[i] == '.'):
                if input_str[i] == '.':
                    if has_dot:
                        raise ValueError(
                            f"Unexpected character '.' at position {i}"
                        )
                    has_dot = True
                i += 1
            tokens.append(Token("number", input_str[start:i]))
            continue

        # Two-character operator: **
        if ch == '*' and i + 1 < n and input_str[i + 1] == '*':
            tokens.append(Token("power", "**"))
            i += 2
            continue

        # Single-character operators and parens
        if ch == '+':
            tokens.append(Token("plus", "+"))
        elif ch == '-':
            tokens.append(Token("minus", "-"))
        elif ch == '*':
            tokens.append(Token("star", "*"))
        elif ch == '/':
            tokens.append(Token("slash", "/"))
        elif ch == '%':
            tokens.append(Token("percent", "%"))
        elif ch == '(':
            tokens.append(Token("lparen", "("))
        elif ch == ')':
            tokens.append(Token("rparen", ")"))
        else:
            raise ValueError(
                f"Unexpected character '{ch}' at position {i}"
            )

        i += 1

    return tokens


# ---------------------------------------------------------------------------
# parser
# ---------------------------------------------------------------------------

def parse(tokens: list[Token]) -> AstNode:
    """Parse a token list into an AST using recursive descent."""
    pos = 0

    def peek() -> Token | None:
        nonlocal pos
        if pos < len(tokens):
            return tokens[pos]
        return None

    def consume() -> Token:
        nonlocal pos
        t = tokens[pos]
        pos += 1
        return t

    def parse_add_sub() -> AstNode:
        node = parse_mul_div()
        while True:
            t = peek()
            if t is None or t.kind not in ("plus", "minus"):
                break
            consume()
            op = "+" if t.kind == "plus" else "-"
            right = parse_mul_div()
            node = BinaryExpr(op, node, right)
        return node

    def parse_mul_div() -> AstNode:
        node = parse_power()
        while True:
            t = peek()
            if t is None or t.kind not in ("star", "slash", "percent"):
                break
            consume()
            if t.kind == "star":
                op = "*"
            elif t.kind == "slash":
                op = "/"
            else:
                op = "%"
            right = parse_power()
            node = BinaryExpr(op, node, right)
        return node

    def parse_power() -> AstNode:
        base = parse_unary()
        t = peek()
        if t is not None and t.kind == "power":
            consume()
            # Right-associative: recurse into parse_power, not parse_unary
            exponent = parse_power()
            return BinaryExpr("**", base, exponent)
        return base

    def parse_unary() -> AstNode:
        t = peek()
        if t is not None and t.kind == "minus":
            consume()
            operand = parse_unary()
            return UnaryExpr("-", operand)
        return parse_atom()

    def parse_atom() -> AstNode:
        t = peek()
        if t is None:
            raise ValueError("Unexpected end of input")

        if t.kind == "number":
            consume()
            return NumberLiteral(float(t.value))

        if t.kind == "lparen":
            consume()
            node = parse_add_sub()
            closing = peek()
            if closing is None or closing.kind != "rparen":
                raise ValueError("Expected rparen")
            consume()
            return node

        raise ValueError(
            f"Unexpected token '{t.kind}' with value '{t.value}'"
        )

    if len(tokens) == 0:
        raise ValueError("Unexpected end of input")

    result = parse_add_sub()

    if pos < len(tokens):
        raise ValueError("Unexpected token after expression")

    return result


# ---------------------------------------------------------------------------
# evaluator
# ---------------------------------------------------------------------------

def evaluate(node: AstNode) -> float:
    """Evaluate an AST node to a numeric result."""
    if isinstance(node, NumberLiteral):
        return node.value

    if isinstance(node, UnaryExpr):
        return -evaluate(node.operand)

    if isinstance(node, BinaryExpr):
        left = evaluate(node.left)
        right = evaluate(node.right)

        if node.op == "+":
            return left + right
        if node.op == "-":
            return left - right
        if node.op == "*":
            return left * right
        if node.op == "/":
            if right == 0:
                raise ValueError("Division by zero")
            return left / right
        if node.op == "%":
            if right == 0:
                raise ValueError("Modulo by zero")
            return left % right
        if node.op == "**":
            return left ** right

    raise ValueError(f"Unknown node type: {type(node)}")


# ---------------------------------------------------------------------------
# evaluate (root: public API)
# ---------------------------------------------------------------------------

def calc(expression: str) -> float:
    """Evaluate a math expression string and return the numeric result."""
    if expression.strip() == "":
        raise ValueError("Empty expression")
    tokens = tokenize(expression)
    ast = parse(tokens)
    return evaluate(ast)

File: research/experiments/mathexpr-ref-go-haiku/mathexpr.go
package main

import (
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
	"unicode"
)

// ============================================================================
// Token types
// ============================================================================

type TokenKind string

const (
	TokenNumber TokenKind = "number"
	TokenPlus   TokenKind = "plus"
	TokenMinus  TokenKind = "minus"
	TokenStar   TokenKind = "star"
	TokenSlash  TokenKind = "slash"
	TokenPercent TokenKind = "percent"
	TokenPower  TokenKind = "power"
	TokenLParen TokenKind = "lparen"
	TokenRParen TokenKind = "rparen"
)

type Token struct {
	Kind  TokenKind
	Value string
}

// ============================================================================
// AST types
// ============================================================================

type BinaryOp string

const (
	OpAdd    BinaryOp = "+"
	OpSub    BinaryOp = "-"
	OpMul    BinaryOp = "*"
	OpDiv    BinaryOp = "/"
	OpMod    BinaryOp = "%"
	OpPower  BinaryOp = "**"
)

type UnaryOp string

const (
	OpNeg UnaryOp = "-"
)

// AstNode is the interface for all AST node types
type AstNode interface {
	isAstNode()
}

// NumberLiteral represents a numeric literal
type NumberLiteral struct {
	Value float64
}

func (n *NumberLiteral) isAstNode() {}

// UnaryExpr represents a unary operation
type UnaryExpr struct {
	Op      UnaryOp
	Operand AstNode
}

func (u *UnaryExpr) isAstNode() {}

// BinaryExpr represents a binary operation
type BinaryExpr struct {
	Op    BinaryOp
	Left  AstNode
	Right AstNode
}

func (b *BinaryExpr) isAstNode() {}

// ============================================================================
// Tokenizer
// ============================================================================

func Tokenize(input string) ([]Token, error) {
	var tokens []Token
	i := 0

	for i < len(input) {
		ch := rune(input[i])

		// Skip whitespace
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		// Parentheses
		if ch == '(' {
			tokens = append(tokens, Token{TokenLParen, "("})
			i++
			continue
		}

		if ch == ')' {
			tokens = append(tokens, Token{TokenRParen, ")"})
			i++
			continue
		}

		// Single-char operators
		if ch == '+' {
			tokens = append(tokens, Token{TokenPlus, "+"})
			i++
			continue
		}

		if ch == '-' {
			tokens = append(tokens, Token{TokenMinus, "-"})
			i++
			continue
		}

		// Power operator
		if ch == '*' {
			if i+1 < len(input) && input[i+1] == '*' {
				tokens = append(tokens, Token{TokenPower, "**"})
				i += 2
			} else {
				tokens = append(tokens, Token{TokenStar, "*"})
				i++
			}
			continue
		}

		if ch == '/' {
			tokens = append(tokens, Token{TokenSlash, "/"})
			i++
			continue
		}

		if ch == '%' {
			tokens = append(tokens, Token{TokenPercent, "%"})
			i++
			continue
		}

		// Numbers
		if unicode.IsDigit(ch) || ch == '.' {
			num := ""
			hasDot := false
			for i < len(input) && (unicode.IsDigit(rune(input[i])) || input[i] == '.') {
				if input[i] == '.' {
					if hasDot {
						return nil, fmt.Errorf("Unexpected character '.' at position %d", i)
					}
					hasDot = true
				}
				num += string(input[i])
				i++
			}
			tokens = append(tokens, Token{TokenNumber, num})
			continue
		}

		return nil, fmt.Errorf("Unexpected character '%c' at position %d", ch, i)
	}

	return tokens, nil
}

// ============================================================================
// Parser
// ============================================================================

type parser struct {
	tokens []Token
	pos    int
}

func (p *parser) peek() *Token {
	if p.pos < len(p.tokens) {
		return &p.tokens[p.pos]
	}
	return nil
}

func (p *parser) advance() Token {
	t := p.tokens[p.pos]
	p.pos++
	return t
}

func (p *parser) expect(kind TokenKind) (Token, error) {
	t := p.peek()
	if t == nil || t.Kind != kind {
		if t == nil {
			return Token{}, fmt.Errorf("Expected %s but got end of input", kind)
		}
		return Token{}, fmt.Errorf("Expected %s but got %s", kind, t.Kind)
	}
	return p.advance(), nil
}

func (p *parser) parseAddSub() (AstNode, error) {
	left, err := p.parseMulDiv()
	if err != nil {
		return nil, err
	}

	for {
		t := p.peek()
		if t == nil || (t.Kind != TokenPlus && t.Kind != TokenMinus) {
			break
		}

		opToken := p.advance()
		var op BinaryOp
		if opToken.Kind == TokenPlus {
			op = OpAdd
		} else {
			op = OpSub
		}

		right, err := p.parseMulDiv()
		if err != nil {
			return nil, err
		}

		left = &BinaryExpr{op, left, right}
	}

	return left, nil
}

func (p *parser) parseMulDiv() (AstNode, error) {
	left, err := p.parsePower()
	if err != nil {
		return nil, err
	}

	for {
		t := p.peek()
		if t == nil || (t.Kind != TokenStar && t.Kind != TokenSlash && t.Kind != TokenPercent) {
			break
		}

		opToken := p.advance()
		var op BinaryOp
		switch opToken.Kind {
		case TokenStar:
			op = OpMul
		case TokenSlash:
			op = OpDiv
		case TokenPercent:
			op = OpMod
		}

		right, err := p.parsePower()
		if err != nil {
			return nil, err
		}

		left = &BinaryExpr{op, left, right}
	}

	return left, nil
}

func (p *parser) parsePower() (AstNode, error) {
	base, err := p.parseUnary()
	if err != nil {
		return nil, err
	}

	t := p.peek()
	if t != nil && t.Kind == TokenPower {
		p.advance()
		exponent, err := p.parsePower() // right-recursive for right-associativity
		if err != nil {
			return nil, err
		}
		return &BinaryExpr{OpPower, base, exponent}, nil
	}

	return base, nil
}

func (p *parser) parseUnary() (AstNode, error) {
	t := p.peek()
	if t != nil && t.Kind == TokenMinus {
		p.advance()
		operand, err := p.parseUnary() // allow chained unary
		if err != nil {
			return nil, err
		}
		return &UnaryExpr{OpNeg, operand}, nil
	}

	return p.parseAtom()
}

func (p *parser) parseAtom() (AstNode, error) {
	t := p.peek()

	if t == nil {
		return nil, errors.New("Unexpected end of input")
	}

	if t.Kind == TokenNumber {
		p.advance()
		val, err := strconv.ParseFloat(t.Value, 64)
		if err != nil {
			return nil, err
		}
		return &NumberLiteral{val}, nil
	}

	if t.Kind == TokenLParen {
		p.advance()
		expr, err := p.parseAddSub()
		if err != nil {
			return nil, err
		}
		_, err = p.expect(TokenRParen)
		if err != nil {
			return nil, err
		}
		return expr, nil
	}

	return nil, fmt.Errorf("Unexpected token: %s '%s'", t.Kind, t.Value)
}

func Parse(tokens []Token) (AstNode, error) {
	p := &parser{tokens, 0}
	ast, err := p.parseAddSub()
	if err != nil {
		return nil, err
	}

	if p.pos < len(p.tokens) {
		remaining := p.tokens[p.pos]
		return nil, fmt.Errorf("Unexpected token after expression: %s '%s'", remaining.Kind, remaining.Value)
	}

	return ast, nil
}

// ============================================================================
// Evaluator
// ============================================================================

func Evaluate(node AstNode) (float64, error) {
	switch n := node.(type) {
	case *NumberLiteral:
		return n.Value, nil

	case *UnaryExpr:
		operand, err := Evaluate(n.Operand)
		if err != nil {
			return 0, err
		}
		return -operand, nil

	case *BinaryExpr:
		left, err := Evaluate(n.Left)
		if err != nil {
			return 0, err
		}
		right, err := Evaluate(n.Right)
		if err != nil {
			return 0, err
		}

		switch n.Op {
		case OpAdd:
			return left + right, nil
		case OpSub:
			return left - right, nil
		case OpMul:
			return left * right, nil
		case OpPower:
			return math.Pow(left, right), nil
		case OpDiv:
			if right == 0 {
				return 0, errors.New("Division by zero")
			}
			return left / right, nil
		case OpMod:
			if right == 0 {
				return 0, errors.New("Modulo by zero")
			}
			// Go's % operator works with floats via int conversion
			// For floating-point modulo, use math.Mod
			return math.Mod(left, right), nil
		default:
			return 0, fmt.Errorf("Unknown operator: %s", n.Op)
		}

	default:
		return 0, errors.New("Unknown AST node type")
	}
}

// ============================================================================
// Public API
// ============================================================================

// Calc evaluates a math expression string and returns the numeric result.
func Calc(expression string) (float64, error) {
	trimmed := strings.TrimSpace(expression)
	if trimmed == "" {
		return 0, errors.New("Empty expression")
	}

	tokens, err := Tokenize(expression)
	if err != nil {
		return 0, err
	}

	ast, err := Parse(tokens)
	if err != nil {
		return 0, err
	}

	return Evaluate(ast)
}
